File: zul/src/com/potix/zul/html/Bandpopup.java
/* Bandpopup.java

{{IS_NOTE
	Purpose:
		
	Description:
		
	History:
		Mon Mar 20 12:31:44     2006, Created by tomyeh@potix.com
}}IS_NOTE

Copyright (C) 2006 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
}}IS_RIGHT
*/
package com.potix.zul.html;

import com.potix.zk.ui.Component;
import com.potix.zk.ui.UiException;
import com.potix.zul.html.impl.XulElement;

/**
 * The popup that belongs to a {@link Bandbox} instance.
 *
 * <p>Developer usually listen to the onOpen event that is sent to
 * {@link Bandbox} and then creates proper components as children
 * of this component.
 *
 * @author <a href="mailto:tomyeh@potix.com">tomyeh@potix.com</a>
 * @version $Revision: 1.3 $ $Date: 2006/05/29 04:28:20 $
 */
public class Bandpopup extends XulElement {
	public Bandpopup() {
	}

	//-- super --//
	public boolean setVisible(boolean visible) {
		if (!visible)
			throw new UnsupportedOperationException("Use Bandbox.setOpen(false) instead");
		return true;
	}
	public void setParent(Component parent) {
		if (parent != null && !(parent instanceof Bandbox))
			throw new UiException("Bandpopup's parent must be Bandbox");
		super.setParent(parent);
	}
}

File: zul/src/archive/web/js/zul/inp/Bandpopup.js
/* Bandpopup.js

	Purpose:

	Description:

	History:
		Fri Apr  3 15:24:37     2009, Created by tomyeh

Copyright (C) 2009 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
/**
 * The popup that belongs to a {@link Bandbox} instance.
 *
 * <p>Developer usually listen to the onOpen event that is sent to
 * {@link Bandbox} and then creates proper components as children
 * of this component.
 *
 * <p>z-class: z-bandpopup
 */
zul.inp.Bandpopup = zk.$extends(zul.Widget, {
	bind_: function () {
		this.$supers(zul.inp.Bandpopup, 'bind_', arguments);
		jq(this.$n()).on('focusin', this.proxy(this._focusin))
			.on('focusout', this.proxy(this._focusout));
	},
	unbind_: function () {
		jq(this.$n()).off('focusout', this.proxy(this._focusout))
			.off('focusin', this.proxy(this._focusin));
		this.$supers(zul.inp.Bandpopup, 'unbind_', arguments);
	},
	_focusin: function (e) {
		this._shallClosePopup = false;
	},
	_focusout: function (e) {
		var bandbox = this.parent,
			self = this;
		if (e.relatedTarget) {
			if (bandbox && bandbox.isOpen() && !jq.isAncestor(this.$n(), e.relatedTarget))
				bandbox.close();
		} else {
			// for solving B96-ZK-4748, treechildren will rerender itself when clicking
			// the open icon, and JQ will simulate a fake focusout event without any relatedTarget.

			self._shallClosePopup = true;
			setTimeout(function () {
				if (bandbox && bandbox.isOpen() && self._shallClosePopup) {
					bandbox.close();
					self._shallClosePopup = false;
				}
			});
		}
	},
	//super
	afterChildrenMinFlex_: function (orient) {
		if (orient == 'w') {
			var bandbox = this.parent,
				pp = bandbox && bandbox.$n('pp');
			if (pp) {
				// test case is B50-ZK-859.zul
				pp.style.width = jq.px0(this._hflexsz + zk(pp).padBorderWidth());
				zk(pp)._updateProp(['width']);
			}
		}
	},
	doClick_: function (evt) {
		if (evt.domTarget == this.$n())
			this.parent.focus();
		this.$supers('doClick_', arguments);
	}
});
